src/core/ExpressionFormatter.ts

```diff
@@ -0,0 +1,383 @@
+    this.layout.add(WS.INDENT);
+    this.layout = this.formatSubExpression(node.children);
+  }
+
+  private formatLimitClause(node: LimitClause) {
+    this.layout.add(WS.NEWLINE, WS.INDENT, this.show(node.limitToken));
+    this.layout.indentation.increaseTopLevel();
+
+    if (node.offsetToken) {
+      this.layout.add(
+        WS.NEWLINE,
+        WS.INDENT,
+        this.show(node.offsetToken),
+        ',',
+        WS.SPACE,
+        this.show(node.countToken),
+        WS.SPACE
+      );
+    } else {
+      this.layout.add(WS.NEWLINE, WS.INDENT, this.show(node.countToken), WS.SPACE);
+    }
+    this.layout.indentation.decreaseTopLevel();
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  private formatAllColumnsAsterisk(node: AllColumnsAsterisk) {
+    this.layout.add('*', WS.SPACE);
+  }
+
+  private formatSubExpression(nodes: AstNode[], inline = this.inline): Layout {
+    return new ExpressionFormatter({
+      cfg: this.cfg,
+      params: this.params,
+      layout: this.layout,
+      inline,
+    }).format(nodes);
+  }
+
+  private formatToken(token: Token): void {
+    switch (token.type) {
+      case TokenType.LINE_COMMENT:
+        return this.formatLineComment(token);
+      case TokenType.BLOCK_COMMENT:
+        return this.formatBlockComment(token);
+      case TokenType.RESERVED_JOIN:
+        return this.formatJoin(token);
+      case TokenType.RESERVED_DEPENDENT_CLAUSE:
+        return this.formatDependentClause(token);
+      case TokenType.RESERVED_JOIN_CONDITION:
+        return this.formatJoinCondition(token);
+      case TokenType.RESERVED_LOGICAL_OPERATOR:
+        return this.formatLogicalOperator(token);
+      case TokenType.RESERVED_KEYWORD:
+        return this.formatKeyword(token);
+      case TokenType.RESERVED_CASE_START:
+        return this.formatCaseStart(token);
+      case TokenType.RESERVED_CASE_END:
+        return this.formatCaseEnd(token);
+      case TokenType.PARAMETER:
+        return this.formatParameter(token);
+      case TokenType.OPERATOR:
+        return this.formatOperator(token);
+      case TokenType.IDENT:
+      case TokenType.STRING:
+      case TokenType.NUMBER:
+      case TokenType.VARIABLE:
+        return this.formatWord(token);
+      default:
+        throw new Error(`Unexpected token type: ${token.type}`);
+    }
+  }
+
+  /**
+   * Formats ident/string/number/variable tokens
+   */
+  private formatWord(token: Token) {
+    this.layout.add(this.show(token), WS.SPACE);
+  }
+
+  /** Formats a line comment onto query */
+  private formatLineComment(token: Token) {
+    this.layout.add(this.show(token), WS.NEWLINE, WS.INDENT);
+  }
+
+  /** Formats a block comment onto query */
+  private formatBlockComment(token: Token) {
+    this.splitBlockComment(token.value).forEach(line => {
+      this.layout.add(WS.NEWLINE, WS.INDENT, line);
+    });
+    this.layout.add(WS.NEWLINE, WS.INDENT);
+  }
+
+  private splitBlockComment(comment: string): string[] {
```

Comment:
add comment for this function

Reply:
Good point. This can be confusing to understand. Added comment.